File: CLIENT/src/controllers/DashboardController.js
// controllers/DashboardController.js
import { API_CONFIG } from "../config/APIConfig.js";
import { apiCall } from "@/helpers/APIHelper.js";
import { apiCallWithTokenRefresh, extractTokenFromHeader } from "./AuthController.js";

// Helper function to get authorization header
const getAuthHeaders = (jwtToken) => ({
  Authorization: `Bearer ${jwtToken}`,
  "Content-Type": "application/json"
});

// Helper to build query parameters
const buildQueryParams = (params = {}) => {
  const queryParams = new URLSearchParams();
  
  Object.keys(params).forEach(key => {
    if (params[key] !== undefined && params[key] !== null && params[key] !== '') {
      queryParams.append(key, params[key]);
    }
  });
  
  return queryParams.toString();
};

// ============ DASHBOARD METHODS ============

/**
 * Get lightweight initial dashboard data for fast initial load
 * @param {string} authorizationHeader - Bearer token
 * @returns {Promise} API response with minimal dashboard data
 */
export const getInitialDashboard = async (authorizationHeader) => {
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(`/dashboard/initial`, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Get dashboard statistics
 * @param {string} authorizationHeader - Bearer token
 * @returns {Promise} API response
 */
export const getDashboardStats = async (authorizationHeader) => {
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(`/dashboard/stats`, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};



/**
 * Get details of a generated API by ID
 * @param {string} authorizationHeader - Bearer token
 * @param {string} apiId - The ID of the generated API
 * @returns {Promise} API response
 */
export const getGeneratedApiDetails = async (authorizationHeader, apiId) => {
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(`/gen/api/${apiId}`, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};


/**
 * Get collections with pagination
 * @param {string} authorizationHeader - Bearer token
 * @param {Object} pagination - Pagination parameters
 * @param {number} pagination.page - Page number (0-based, default: 0)
 * @param {number} pagination.size - Page size (default: 10)
 * @param {string} pagination.sortBy - Sort field (default: 'name')
 * @param {string} pagination.sortDir - Sort direction 'asc' or 'desc' (default: 'asc')
 * @returns {Promise} API response
 */
export const getDashboardCollections = async (authorizationHeader, pagination = {}) => {
  const { page = 0, size = 10, sortBy = 'name', sortDir = 'asc' } = pagination;
  
  const queryString = buildQueryParams({ page, size, sortBy, sortDir });
  const url = `/dashboard/collections${queryString ? `?${queryString}` : ''}`;
  
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(url, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Get endpoints with pagination and filtering
 * @param {string} authorizationHeader - Bearer token
 * @param {Object} params - Query parameters
 * @param {number} params.page - Page number (0-based, default: 0)
 * @param {number} params.size - Page size (default: 10)
 * @param {string} params.collectionId - Filter by collection ID
 * @param {string} params.method - Filter by HTTP method
 * @param {string} params.search - Search term
 * @param {string} params.sortBy - Sort field (default: 'lastUpdated')
 * @param {string} params.sortDir - Sort direction (default: 'desc')
 * @returns {Promise} API response
 */
export const getDashboardEndpoints = async (authorizationHeader, params = {}) => {
  const { 
    page = 0, 
    size = 10, 
    collectionId, 
    method, 
    search,
    sortBy = 'lastUpdated', 
    sortDir = 'desc' 
  } = params;
  
  const queryString = buildQueryParams({ 
    page, 
    size, 
    collectionId, 
    method, 
    search,
    sortBy, 
    sortDir 
  });
  
  const url = `/dashboard/endpoints${queryString ? `?${queryString}` : ''}`;
  
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(url, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Get recent activities with pagination
 * @param {string} authorizationHeader - Bearer token
 * @param {Object} params - Query parameters
 * @param {number} params.page - Page number (0-based, default: 0)
 * @param {number} params.size - Page size (default: 20)
 * @param {string} params.from - Start date (ISO datetime)
 * @param {string} params.to - End date (ISO datetime)
 * @returns {Promise} API response
 */
export const getDashboardActivities = async (authorizationHeader, params = {}) => {
  const { page = 0, size = 20, from, to } = params;
  
  const queryString = buildQueryParams({ page, size, from, to });
  const url = `/dashboard/activities${queryString ? `?${queryString}` : ''}`;
  
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(url, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Get security summary
 * @param {string} authorizationHeader - Bearer token
 * @returns {Promise} API response
 */
export const getDashboardSecuritySummary = async (authorizationHeader) => {
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(`/dashboard/security/summary`, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Get supported programming languages
 * @param {string} authorizationHeader - Bearer token
 * @returns {Promise} API response
 */
export const getDashboardLanguages = async (authorizationHeader) => {
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(`/dashboard/languages`, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Get environments overview
 * @param {string} authorizationHeader - Bearer token
 * @returns {Promise} API response
 */
export const getDashboardEnvironments = async (authorizationHeader) => {
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(`/dashboard/environments`, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

/**
 * Global search across dashboard
 * @param {string} authorizationHeader - Bearer token
 * @param {Object} searchParams - Search parameters
 * @param {string} searchParams.q - Search query (required)
 * @param {string} searchParams.types - Comma-separated types to search (collections,endpoints,users)
 * @param {number} searchParams.page - Page number (0-based, default: 0)
 * @param {number} searchParams.size - Page size (default: 20)
 * @returns {Promise} API response
 */
export const globalDashboardSearch = async (authorizationHeader, searchParams = {}) => {
  const { q, types, page = 0, size = 20 } = searchParams;
  
  if (!q) {
    throw new Error('Search query (q) is required');
  }
  
  const queryString = buildQueryParams({ q, types, page, size });
  const url = `/dashboard/search${queryString ? `?${queryString}` : ''}`;
  
  return apiCallWithTokenRefresh(
    authorizationHeader,
    (authHeader) => apiCall(url, {
      method: 'GET',
      headers: getAuthHeaders(authHeader.replace('Bearer ', ''))
    })
  );
};

// ============ RESPONSE HANDLERS & UTILITIES ============

/**
 * Handle standardized API responses for dashboard operations
 * @param {Object} response - API response
 * @returns {Object} Processed response data
 */
export const handleDashboardResponse = (response) => {
  if (!response) {
    throw new Error('No response received from dashboard service');
  }

  // The Java controller returns responses with responseCode field
  const responseCode = response.responseCode || response.status;
  
  if (responseCode === 200) {
    return {
      ...response,
      data: response.data || {},
      responseCode: responseCode,
      requestId: response.requestId,
      timestamp: response.timestamp
    };
  }

  const errorMessage = response.message || response.error || 'Unknown error';
  switch (responseCode) {
    case 400: 
      throw new Error(`Bad Request: ${errorMessage}`);
    case 401: 
      throw new Error(`Unauthorized: ${errorMessage}`);
    case 403: 
      throw new Error(`Forbidden: ${errorMessage}`);
    case 404: 
      throw new Error(`Not Found: ${errorMessage}`);
    case 500: 
      throw new Error(`Server Error: ${errorMessage}`);
    default: 
      throw new Error(`Error ${responseCode || 'Unknown'}: ${errorMessage}`);
  }
};

/**
 * Extract paginated response data
 * @param {Object} response - API response
 * @returns {Object} Paginated data structure
 */
export const extractPaginatedData = (response) => {
  if (!response || !response.data) return null;
  
  return {
    content: response.data.content || [],
    pageNumber: response.data.pageNumber || 0,
    pageSize: response.data.pageSize || 0,
    totalElements: response.data.totalElements || 0,
    totalPages: response.data.totalPages || 0,
    last: response.data.last || true,
    requestId: response.requestId
  };
};

/**
 * Validate dashboard search parameters
 * @param {Object} searchParams - Search parameters to validate
 * @returns {Array} Array of validation errors
 */
export const validateDashboardSearchParams = (searchParams) => {
  const errors = [];
  
  if (!searchParams.q || searchParams.q.trim() === '') {
    errors.push('Search query is required');
  }
  
  if (searchParams.page !== undefined && (searchParams.page < 0 || !Number.isInteger(searchParams.page))) {
    errors.push('Page must be a non-negative integer');
  }
  
  if (searchParams.size !== undefined && (searchParams.size < 1 || searchParams.size > 100)) {
    errors.push('Size must be between 1 and 100');
  }
  
  if (searchParams.types) {
    const validTypes = ['collections', 'endpoints', 'users'];
    const types = searchParams.types.split(',');
    const invalidTypes = types.filter(type => !validTypes.includes(type));
    if (invalidTypes.length > 0) {
      errors.push(`Invalid types: ${invalidTypes.join(', ')}. Valid types: ${validTypes.join(', ')}`);
    }
  }
  
  return errors;
};

/**
 * Format date for display
 * @param {string|Date} date - Date to format
 * @returns {string} Formatted date string
 */
export const formatDateForDisplay = (date) => {
  if (!date) return '';
  
  const d = new Date(date);
  return d.toLocaleString('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit'
  });
};

/**
 * Refresh initial dashboard data (lightweight)
 * @param {string} authorizationHeader - Bearer token
 * @returns {Promise} Initial dashboard data
 */
export const refreshInitialDashboard = async (authorizationHeader) => {
  try {
    const response = await getInitialDashboard(authorizationHeader);
    return handleDashboardResponse(response);
  } catch (error) {
    throw new Error(`Failed to refresh initial dashboard: ${error.message}`);
  }
};

/**
 * Format dashboard data for display
 * @param {Object} dashboardData - Raw dashboard data
 * @returns {Object} Formatted dashboard data with date strings
 */
export const formatDashboardData = (dashboardData) => {
  if (!dashboardData) return {};
  
  const formatted = { ...dashboardData };
  
  // Format timestamps in the response
  if (formatted.timestamp) {
    formatted.formattedTimestamp = formatDateForDisplay(formatted.timestamp);
  }
  
  return formatted;
};

// Export all methods with consistent naming
export default {
  getInitialDashboard,
  getDashboardStats,
  getDashboardCollections,
  getDashboardEndpoints,
  getDashboardActivities,
  getDashboardSecuritySummary,
  getDashboardLanguages,
  getDashboardEnvironments,
  globalDashboardSearch,
  handleDashboardResponse,
  extractPaginatedData,
  validateDashboardSearchParams,
  formatDateForDisplay,
  refreshInitialDashboard,
  formatDashboardData
};

File: SERVER/src/main/java/com/usg/apiGeneration/controllers/DashboardController.java
package com.usg.apiGeneration.controllers;

import com.usg.apiGeneration.dtos.dashboard.*;
import com.usg.apiGeneration.helpers.JwtHelper;
import com.usg.apiGeneration.services.DashboardService;
import com.usg.apiGeneration.utils.LoggerUtil;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.enums.ParameterIn;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Slf4j
@RestController
@RequestMapping("/plx/api/dashboard")
@RequiredArgsConstructor
@Tag(name = "DASHBOARD", description = "Endpoints for dashboard statistics and monitoring")
public class DashboardController {

    private final DashboardService dashboardService;
    private final LoggerUtil loggerUtil;
    private final JwtHelper jwtHelper;

    private ResponseEntity<?> validateAuth(HttpServletRequest req, String action, String requestId) {
        ResponseEntity<?> authValidation = jwtHelper.validateAuthorizationHeader(req, action);
        if (authValidation != null) {
            loggerUtil.log("dashboard", "Request ID: " + requestId + ", Authorization failed for " + action);
        }
        return authValidation;
    }

    private Map<String, Object> createSuccessResponse(String message, Object data, String requestId) {
        Map<String, Object> response = new HashMap<>();
        response.put("responseCode", 200);
        response.put("message", message);
        response.put("data", data);
        response.put("requestId", requestId);
        response.put("timestamp", LocalDateTime.now().toString());
        return response;
    }

    private Map<String, Object> createErrorResponse(String message, String requestId, HttpStatus status) {
        Map<String, Object> response = new HashMap<>();
        response.put("responseCode", status.value());
        response.put("message", message);
        response.put("requestId", requestId);
        response.put("timestamp", LocalDateTime.now().toString());
        return response;
    }

    // ============================================================
    // 1. DASHBOARD STATS (Fast, lightweight)
    // ============================================================
    @GetMapping("/stats")
    @Operation(summary = "Get dashboard statistics", parameters = {
            @Parameter(name = "Authorization", description = "JWT Token", required = true, in = ParameterIn.HEADER)
    })
    public ResponseEntity<?> getDashboardStats(HttpServletRequest req) {
        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting dashboard statistics", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            DashboardStatsResponseDTO stats = dashboardService.getDashboardStats(requestId, req, performedBy);

            loggerUtil.log("dashboard", "Request ID: " + requestId + ", Stats retrieved successfully");
            return ResponseEntity.ok(createSuccessResponse("Dashboard statistics retrieved successfully", stats, requestId));
        } catch (Exception e) {
            loggerUtil.log("dashboard", "Request ID: " + requestId + ", Error: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting statistics: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }

    // ============================================================
    // 2. COLLECTIONS with Pagination
    // ============================================================
    @GetMapping("/collections")
    @Operation(summary = "Get collections with pagination", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER),
            @Parameter(name = "page", description = "Page number (0-based)", in = ParameterIn.QUERY),
            @Parameter(name = "size", description = "Page size", in = ParameterIn.QUERY),
            @Parameter(name = "sortBy", description = "Sort field", in = ParameterIn.QUERY),
            @Parameter(name = "sortDir", description = "Sort direction", in = ParameterIn.QUERY)
    })
    public ResponseEntity<?> getCollections(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size,
            @RequestParam(defaultValue = "name") String sortBy,
            @RequestParam(defaultValue = "asc") String sortDir,
            HttpServletRequest req) {

        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting collections", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            Sort.Direction direction = sortDir.equalsIgnoreCase("desc") ? Sort.Direction.DESC : Sort.Direction.ASC;
            Pageable pageable = PageRequest.of(page, size, Sort.by(direction, sortBy));

            PaginatedResponseDTO<DashboardCollectionDTO> collections =
                    dashboardService.getCollectionsPaginated(requestId, req, performedBy, pageable);

            return ResponseEntity.ok(createSuccessResponse("Collections retrieved successfully", collections, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting collections: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }

    // ============================================================
    // 3. ENDPOINTS with Pagination and Filtering
    // ============================================================
    @GetMapping("/endpoints")
    @Operation(summary = "Get endpoints with pagination and filtering", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER),
            @Parameter(name = "page", description = "Page number (0-based)", in = ParameterIn.QUERY),
            @Parameter(name = "size", description = "Page size", in = ParameterIn.QUERY),
            @Parameter(name = "collectionId", description = "Filter by collection", in = ParameterIn.QUERY),
            @Parameter(name = "method", description = "Filter by HTTP method", in = ParameterIn.QUERY),
            @Parameter(name = "search", description = "Search term", in = ParameterIn.QUERY),
            @Parameter(name = "sortBy", description = "Sort field (lastUpdated, createdAt, updatedAt, name, method, etc.)", in = ParameterIn.QUERY),
            @Parameter(name = "sortDir", description = "Sort direction (asc/desc)", in = ParameterIn.QUERY)
    })
    public ResponseEntity<?> getEndpoints(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size,
            @RequestParam(required = false) String collectionId,
            @RequestParam(required = false) String method,
            @RequestParam(required = false) String search,
            @RequestParam(defaultValue = "lastUpdated") String sortBy,
            @RequestParam(defaultValue = "desc") String sortDir,
            HttpServletRequest req) {

        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting endpoints", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);

            // Create filter DTO
            EndpointFilterDTO filter = EndpointFilterDTO.builder()
                    .collectionId(collectionId)
                    .method(method)
                    .search(search)
                    .build();

            // Get paginated endpoints with sorting
            PaginatedResponseDTO<DashboardEndpointDTO> endpoints =
                    dashboardService.getEndpointsPaginated(
                            requestId, req, performedBy, filter,
                            page, size, sortBy, sortDir);

            return ResponseEntity.ok(createSuccessResponse("Endpoints retrieved successfully", endpoints, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting endpoints: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }

    // ============================================================
    // 4. RECENT ACTIVITIES with Pagination
    // ============================================================
    @GetMapping("/activities")
    @Operation(summary = "Get recent activities with pagination", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER),
            @Parameter(name = "page", description = "Page number (0-based)", in = ParameterIn.QUERY),
            @Parameter(name = "size", description = "Page size", in = ParameterIn.QUERY),
            @Parameter(name = "from", description = "Start date", in = ParameterIn.QUERY),
            @Parameter(name = "to", description = "End date", in = ParameterIn.QUERY)
    })
    public ResponseEntity<?> getActivities(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime to,
            HttpServletRequest req) {

        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting activities", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "timestamp"));

            PaginatedResponseDTO<ActivityDTO> activities =
                    dashboardService.getActivitiesPaginated(requestId, req, performedBy, from, to, pageable);

            return ResponseEntity.ok(createSuccessResponse("Activities retrieved successfully", activities, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting activities: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }

    // ============================================================
    // 5. SECURITY SUMMARY
    // ============================================================
    @GetMapping("/security/summary")
    @Operation(summary = "Get security summary", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER)
    })
    public ResponseEntity<?> getSecuritySummary(HttpServletRequest req) {
        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting security summary", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            DashboardSecuritySummaryResponseDTO summary = dashboardService.getDashboardSecuritySummary(requestId, req, performedBy);

            return ResponseEntity.ok(createSuccessResponse("Security summary retrieved successfully", summary, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting security summary: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }



    // ============================================================
    // 11. LANGUAGES (Code Generation)
    // ============================================================
    @GetMapping("/languages")
    @Operation(summary = "Get supported languages", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER)
    })
    public ResponseEntity<?> getLanguages(HttpServletRequest req) {
        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting languages", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            DashboardLanguagesResponseDTO languages = dashboardService.getDashboardLanguages(requestId, req, performedBy);

            return ResponseEntity.ok(createSuccessResponse("Languages retrieved successfully", languages, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting languages: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }


    // ============================================================
    // 13. ENVIRONMENTS
    // ============================================================
    @GetMapping("/environments")
    @Operation(summary = "Get environments", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER)
    })
    public ResponseEntity<?> getEnvironments(HttpServletRequest req) {
        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting environments", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            DashboardEnvironmentsResponseDTO environments = dashboardService.getDashboardEnvironments(requestId, req, performedBy);

            return ResponseEntity.ok(createSuccessResponse("Environments retrieved successfully", environments, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting environments: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }


    // ============================================================
    // 16. SEARCH (Global search across all sections)
    // ============================================================
    @GetMapping("/search")
    @Operation(summary = "Global search across dashboard", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER),
            @Parameter(name = "q", description = "Search query", required = true, in = ParameterIn.QUERY),
            @Parameter(name = "types", description = "Comma-separated types to search (collections,endpoints,users)", in = ParameterIn.QUERY),
            @Parameter(name = "page", in = ParameterIn.QUERY),
            @Parameter(name = "size", in = ParameterIn.QUERY)
    })
    public ResponseEntity<?> globalSearch(
            @RequestParam("q") String query,
            @RequestParam(required = false) String types,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            HttpServletRequest req) {

        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "global search", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);
            Pageable pageable = PageRequest.of(page, size);

            SearchResponseDTO searchResults = dashboardService.globalSearch(
                    requestId, req, performedBy, query, types, pageable);

            return ResponseEntity.ok(createSuccessResponse("Search completed successfully", searchResults, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error during search: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }

    // ============================================================
    // 17. LIGHTWEIGHT INITIAL DASHBOARD (Fast initial load)
    // ============================================================
    @GetMapping("/initial")
    @Operation(summary = "Get lightweight initial dashboard data", parameters = {
            @Parameter(name = "Authorization", required = true, in = ParameterIn.HEADER)
    })
    public ResponseEntity<?> getInitialDashboard(HttpServletRequest req) {
        String requestId = UUID.randomUUID().toString();

        ResponseEntity<?> authValidation = validateAuth(req, "getting initial dashboard", requestId);
        if (authValidation != null) return authValidation;

        try {
            String performedBy = jwtHelper.extractPerformedBy(req);

            // Only fetch essential data for initial render
            InitialDashboardResponseDTO initialData = InitialDashboardResponseDTO.builder()
                    .stats(dashboardService.getDashboardStats(requestId, req, performedBy))
                    .recentCollections(dashboardService.getRecentCollections(requestId, req, performedBy, 3))
                    .recentEndpoints(dashboardService.getRecentEndpoints(requestId, req, performedBy, 5))
                    .recentActivities(dashboardService.getRecentActivities(requestId, req, performedBy, 5))
                    .securitySummary(dashboardService.getDashboardSecuritySummary(requestId, req, performedBy))
                    .build();

            return ResponseEntity.ok(createSuccessResponse("Initial dashboard data retrieved successfully", initialData, requestId));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(createErrorResponse("Error getting initial dashboard: " + e.getMessage(), requestId, HttpStatus.INTERNAL_SERVER_ERROR));
        }
    }
}